Mathlib/Topology/Instances/AddCircle.lean

```diff
@@ -0,0 +1,667 @@
+  refine' ⟨_, fun m hn h0 => _⟩ <;> simp only [Ne, ← coe_nsmul, nsmul_eq_mul]
+  · rw [mul_div_cancel' _ h.ne', coe_period]
+  rw [coe_eq_zero_of_pos_iff p hp.out (mul_pos (Nat.cast_pos.2 h0) <| div_pos hp.out h)]
+  rintro ⟨k, hk⟩
+  rw [mul_div, eq_div_iff h.ne', nsmul_eq_mul, mul_right_comm, ← Nat.cast_mul,
+    (mul_left_injective₀ hp.out.ne').eq_iff, Nat.cast_inj, mul_comm] at hk
+  exact (Nat.le_of_dvd h0 ⟨_, hk.symm⟩).not_lt hn
+#align add_circle.add_order_of_period_div AddCircle.addOrderOf_period_div
+
+variable (p)
+
+theorem gcd_mul_addOrderOf_div_eq {n : ℕ} (m : ℕ) (hn : 0 < n) :
+    m.gcd n * addOrderOf (↑(↑m / ↑n * p) : AddCircle p) = n := by
+  rw [mul_comm_div, ← nsmul_eq_mul, coe_nsmul, addOrderOf_nsmul'']
+  · rw [addOrderOf_period_div hn, Nat.gcd_comm, Nat.mul_div_cancel']
+    exact n.gcd_dvd_left m
+  · rw [← addOrderOf_pos_iff, addOrderOf_period_div hn]
+    exact hn
+#align add_circle.gcd_mul_add_order_of_div_eq AddCircle.gcd_mul_addOrderOf_div_eq
+
+variable {p}
+
+theorem addOrderOf_div_of_gcd_eq_one {m n : ℕ} (hn : 0 < n) (h : m.gcd n = 1) :
+    addOrderOf (↑(↑m / ↑n * p) : AddCircle p) = n := by
+  convert gcd_mul_addOrderOf_div_eq p m hn
+  rw [h, one_mul]
+#align add_circle.add_order_of_div_of_gcd_eq_one AddCircle.addOrderOf_div_of_gcd_eq_one
+
+theorem addOrderOf_div_of_gcd_eq_one' {m : ℤ} {n : ℕ} (hn : 0 < n) (h : m.natAbs.gcd n = 1) :
+    addOrderOf (↑(↑m / ↑n * p) : AddCircle p) = n := by
+  induction m
+  · simp only [Int.ofNat_eq_coe, Int.cast_ofNat, Int.natAbs_ofNat] at h⊢
+    exact addOrderOf_div_of_gcd_eq_one hn h
+  · simp only [Int.cast_negSucc, neg_div, neg_mul, coe_neg, addOrderOf_neg]
+    exact addOrderOf_div_of_gcd_eq_one hn h
+#align add_circle.add_order_of_div_of_gcd_eq_one' AddCircle.addOrderOf_div_of_gcd_eq_one'
+
+theorem addOrderOf_coe_rat {q : ℚ} : addOrderOf (↑(↑q * p) : AddCircle p) = q.den := by
+  have : (↑(q.den : ℤ) : 𝕜) ≠ 0 := by
+    norm_cast
+    exact q.pos.ne.symm
+  rw [← @Rat.num_den q, Rat.cast_mk_of_ne_zero _ _ this, Int.cast_ofNat, Rat.num_den,
+    addOrderOf_div_of_gcd_eq_one' q.pos q.reduced]
+#align add_circle.add_order_of_coe_rat AddCircle.addOrderOf_coe_rat
+
+theorem addOrderOf_eq_pos_iff {u : AddCircle p} {n : ℕ} (h : 0 < n) :
+    addOrderOf u = n ↔ ∃ m < n, m.gcd n = 1 ∧ ↑(↑m / ↑n * p) = u := by
+  refine' ⟨QuotientAddGroup.induction_on' u fun k hk => _, _⟩
+  · rintro ⟨m, h₀, h₁, rfl⟩
+    exact addOrderOf_div_of_gcd_eq_one h h₁
+  have h0 := addOrderOf_nsmul_eq_zero (k : AddCircle p)
+  rw [hk, ← coe_nsmul, coe_eq_zero_iff] at h0
+  obtain ⟨a, ha⟩ := h0
+  have h0 : (_ : 𝕜) ≠ 0 := Nat.cast_ne_zero.2 h.ne'
+  rw [nsmul_eq_mul, mul_comm, ← div_eq_iff h0, ← a.div_add_mod' n, add_smul, add_div, zsmul_eq_mul,
+    Int.cast_mul, Int.cast_ofNat, mul_assoc, ← mul_div, mul_comm _ p, mul_div_cancel p h0] at ha
+  have han : _ = a % n := Int.toNat_of_nonneg (Int.emod_nonneg _ <| by exact_mod_cast h.ne')
+  have he := _; refine' ⟨(a % n).toNat, _, _, he⟩
```

Comment:
I think it's worth asking on Zulip what the Lean4 way to spell this is. This was a useful trick, and it would be a shame to lose it.